---
title: "Step3.1-CellRanger-to-Seurat-QC"
author: "Zhang Yongchao"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Seurat - Guided Clustering Tutorial

see: https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html

```{r}
suppressMessages(library(dplyr)) 
suppressMessages(library(Seurat))
suppressMessages(library(patchwork))
suppressMessages(library(ggplot2))
suppressMessages(library(Matrix))
suppressMessages(library(clustree))
suppressMessages(library(scater))
```


## import data

```{r}
cellranger.outputs <- paste(list.dirs("../../../Cellranger-Outputs",recursive = FALSE),"/outs/filtered_feature_bc_matrix",sep = "")
cellranger.outputs
```

```{r}
obj.list <- lapply(cellranger.outputs,function(x){
  project.name <- basename(gsub("/outs/filtered_feature_bc_matrix","",x))
  #message("**** Processing sample: ",project.name)
  cellranger_mat <- Read10X(x)
  seurat.obj <- CreateSeuratObject(counts = cellranger_mat, project = project.name)
  #message("Done!")
  return(seurat.obj)
})
# obj.list
```

## QC

```{r}
obj.list <- lapply(obj.list, function(pbmc){
  # message("* Processing sample: ",pbmc@project.name)
  pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^mt-")
  pbmc[["UMI.Non.mt"]] <- pbmc[["nCount_RNA"]]*(1-pbmc[["percent.mt"]]/100)
  pbmc[["log10GenesPerUMI"]] <- log10(pbmc$nFeature_RNA) / log10(pbmc$nCount_RNA)
  return(pbmc)
})
```

```{r}
raw.merged <- merge(obj.list[[1]],obj.list[-1])
table(raw.merged@meta.data$orig.ident)
```

> [hbctraining - 04_SC_quality_control](https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md)

**number of genes detected per UMI**: this metric with give us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)


```{r fig.width=8,fig.height=10}
VlnPlot(raw.merged,features = "nFeature_RNA", ncol = 1,pt.size = 0) + NoLegend()
VlnPlot(raw.merged,features = "nFeature_RNA", ncol = 1,pt.size = 0,y.max = 9e3) + NoLegend()
VlnPlot(raw.merged,features = "nCount_RNA", ncol = 1,pt.size = 0) + NoLegend()
VlnPlot(raw.merged,features = "nCount_RNA", ncol = 1,pt.size = 0,y.max = 1e5) + NoLegend()
VlnPlot(raw.merged,features = "nCount_RNA", ncol = 1,pt.size = 0,y.max = 2e4) + NoLegend()
VlnPlot(raw.merged,features = "percent.mt", ncol = 1,pt.size = 0) + NoLegend()
VlnPlot(raw.merged,features = "percent.mt", ncol = 1,pt.size = 0,y.max = 30) + NoLegend()
VlnPlot(raw.merged,features = "log10GenesPerUMI", ncol = 1,pt.size = 0) + NoLegend()
VlnPlot(raw.merged,features = "UMI.Non.mt", ncol = 1,pt.size = 0) + NoLegend()
VlnPlot(raw.merged,features = "UMI.Non.mt", ncol = 1,pt.size = 0,y.max = 1e5) + NoLegend()
plot(raw.merged$log10GenesPerUMI,raw.merged$nCount_RNA,pch = 16,col = rgb(126, 126, 126, 10, maxColorValue=255))
```

```{r fig.width=20,fig.height=8}
metadata <- raw.merged@meta.data
  
# Visualize the number UMIs/transcripts per cell
metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)

# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)

# Visualize the distribution of mitochondrial gene expression detected per cell
metadata %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)
```
```{r fig.width=16,fig.height=16}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~orig.ident)
```


```{r}
hist(raw.merged@meta.data$nCount_RNA,breaks = 300,xlim = c(0,120000))
hist(raw.merged@meta.data$UMI.Non.mt,breaks = 300,xlim = c(0,20000))
hist(raw.merged@meta.data$nFeature_RNA,breaks = 70,xlim = c(0,9000))
hist(raw.merged@meta.data$nFeature_RNA,breaks = 100,xlim = c(0,1000))
hist(raw.merged@meta.data$percent.mt,breaks = 60)
rm("raw.merged")
```


```{r fig.width=18,fig.height=6}
for (sample.no in 1:length(obj.list)) {
  x <- obj.list[[sample.no]]
  print(paste("*Processing sample: ",x@project.name,sep = ""))
  plot1 <- FeatureScatter(x, feature1 = "nCount_RNA", feature2 = "percent.mt")
  plot2 <- FeatureScatter(x, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
  print(plot1 + plot2)
}
```


```{r fig.width=10,fig.height=10}
for (sample.no in 1:length(obj.list)) {
  x <- obj.list[[sample.no]]
  print(paste("*Processing sample: ",x@project.name,sep = ""))
  print(plot(x$log10GenesPerUMI,x$nCount_RNA,pch = 16,col = rgb(126, 126, 126, 10, maxColorValue=255)))
  print(plotHighestExprs(as.SingleCellExperiment(x)))
}
```

## save a demo QC matrix

```{r}
cut.var <- c("nFeature_RNA", "nCount_RNA", "percent.mt","log10GenesPerUMI")
cut.limit <- c("low","high")
cut.columns <- paste(rep(cut.var,each = 2),cut.limit,sep = ".")

qc.df <- as.data.frame(matrix(0,ncol = length(cut.columns),
                     nrow = length(cellranger.outputs),
                     dimnames = list(basename(gsub("/outs/filtered_feature_bc_matrix","",cellranger.outputs)),cut.columns)))
qc.df$log10GenesPerUMI.high <- 1
qc.df$nFeature_RNA.low <- 500
qc.df$nCount_RNA.low <- 3000
```

```{r}
qc.mat <- "./Sample.QC.Cutoffs.csv"

if ( file.exists(qc.mat) ) {
  warning("Found old QC matrix!")
  qc.mat.old <- read.csv(qc.mat,stringsAsFactors = FALSE,row.names = 1)
  if ( all(rownames(qc.mat.old) %in% rownames(qc.df)) ) {
    message("\nNo new samples added!")
  }else{
    qc.mat.old <- rbind(qc.mat.old,qc.df[rownames(qc.df)[!rownames(qc.df) %in% rownames(qc.mat.old)],])
    write.csv(qc.mat.old,file = qc.mat,row.names = FALSE)
  }
}else{
  write.csv(qc.df,file = qc.mat,row.names = TRUE)
}
```

## save raw rds

```{r}
Raw.dir <- "./1.Raw_rds/"
if ( !dir.exists(Raw.dir) ) {
 dir.create(Raw.dir)
}

for (sample.no in 1:length(obj.list)) {
  x <- obj.list[[sample.no]]
  # message(paste("*Processing sample: ",x@project.name,sep = ""))
  sampleName <- paste(x@project.name,".rds",sep = "")
  saveRDS(x,file = paste(Raw.dir,sampleName,sep = ""))
}
```

