---
title: "DESeq2"
author: "Zhang Yongchao"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

``` {r}
### Attention!
######*************************
######ENTREZID == NCBI GENE ID
######*************************
###### This suitable for two groups with three replicates for each
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(warn =-1)
```

# RNAseq pipeline

> refer to: https://github.com/twbattaglia/RNAseq-workflow

```{r}
# Basic Parameters
species <- "drosophlia" # or human, mouse
control <- "N1099G"
case <- "HNF4OE"
```

```{r}
kegg.organism <- switch(species,
                        drosophlia = "dme",
                        human = "hsa",
                        mouse = "mmu")
org.db <- switch(species,
                 drosophlia = "org.Dm.eg.db",
                 human = "org.Hs.eg.db",
                 mouse = "org.Mm.eg.db")
```


## Step 7. Importing Gene Counts into R/RStudio

### 7a. load required R-libraries

```{r}
suppressMessages(library(DESeq2))
suppressMessages(library(ggplot2))
suppressMessages(library(pheatmap))
suppressMessages(library(RColorBrewer))
suppressMessages(library(dplyr))
suppressMessages(library(clusterProfiler))
switch (species,
  drosophlia = suppressMessages(library(org.Dm.eg.db)),
  human = suppressMessages(library(org.Hs.eg.db)),
  mouse = suppressMessages(library(org.Mm.eg.db))
)
suppressMessages(library(pathview))
suppressMessages(library(filesstrings))
```

### 7b. Import featureCounts output

```{r}
# Import gene counts table
# skip first row (general command info)
# make row names the gene identifiers
countdata <- read.table("../../6_Counts_featureCounts/final_counts.txt", header = TRUE, skip = 1, row.names = 1)
head(countdata)
```

```{r}
# Remove .bam + '..' from column identifiers
colnames(countdata) <- gsub(".Aligned.sortedByCoord.out.bam", "", colnames(countdata), fixed = T)
colnames(countdata) <- gsub("..", "", colnames(countdata), fixed = T)
# colnames(countdata) <- unlist(lapply(strsplit(colnames(countdata),split = "_"),function(x)paste(x[1:3],collapse = "_")))
```

```{r}
# Remove length/char columns
countdata <- countdata[ ,c(-1:-5)]
```

```{r}
# filter samples
cat(paste("\"",colnames(countdata),"\"",sep="",collapse = ","))
countdata <- countdata[,c("N1099G.Rep3","N1099G.Rep2","N1099G.Rep1","HNF4.OE.Rep3","HNF4.OE.Rep2","HNF4.OE.Rep1")]
```




### 7c. Import or make metadata text file. The SampleID's must be the first column.

```{r}
# Make metadata
colnames(countdata)
metadata <- data.frame(row.names = colnames(countdata),
                       Group = factor(c(rep(control,3),rep(case,3)),levels = c(control,case)), # control should be in front
                       Replicate = c(paste("Rep",1:3,sep = ""),paste("Rep",1:3,sep="")),
                       sampleid = colnames(countdata))
metadata
```

### 7d. Make DESeq2 object from counts and metadata

```{r}
# - countData : count dataframe
# - colData : sample metadata in the dataframe with row names as sampleID's
# - design : The design of the comparisons to use. 
#            Use (~) before the name of the column variable to compare
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = metadata,
                                 design = ~Group)
# Find differential expressed genes
ddsMat <- DESeq(ddsMat)
```

### 7e. Get basic statisics about the number of significant genes

```{r}
# Get results from testing with FDR adjust pvalues
results <- results(ddsMat, pAdjustMethod = "fdr", alpha = 0.05)

# Generate summary of testing. 
summary(results)
```

```{r}
# Save DESeq2 results
saveRDS(ddsMat, file = "ddsMat.rds")
saveRDS(results,file = "results.rds")
# Check directionality of the log2 fold changes
## Log2 fold change is set as (LoGlu / HiGlu)
## Postive fold changes = Increased in LoGlu
## Negative fold changes = Decreased in LoGlu
mcols(results, use.names = T)
```

## Step 8. Annotate gene symbols

### 8a. Gather gene annotation information

```{r}
# Add gene full name
results$description <- mapIds(x = get(org.db),
                              keys = row.names(results),
                              column = "GENENAME",
                              keytype = "SYMBOL",
                              multiVals = "first")
# Add gene symbol
results$symbol <- row.names(results)

# Add ENTREZ ID
results$entrez <- mapIds(x = get(org.db),
                         keys = row.names(results),
                         column = "ENTREZID",
                         keytype = "SYMBOL",
                         multiVals = "first")

# Add ENSEMBL
results$ensembl <- mapIds(x = get(org.db),
                          keys = row.names(results),
                          column = "ENSEMBL",
                          keytype = "SYMBOL",
                          multiVals = "first")

# Subset for only significant genes (q < 0.05)
results_sig <- subset(results, padj < 0.05)
head(results_sig)
```

### 8b. Write all the important results to .txt files

```{r}
if (!dir.exists("counts")) {
  dir.create("./counts")
}
# Write normalized gene counts to a .txt file
write.table(x = as.data.frame(counts(ddsMat,normalized = T)), 
            file = './counts/normalized_counts.txt', 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant normalized gene counts to a .txt file
write.table(x = counts(ddsMat[row.names(results_sig)], normalized = T), 
            file = './counts/normalized_counts_significant.txt', 
            sep = '\t', 
            quote = F, 
            col.names = NA)

# Write the annotated results table to a .txt file
write.table(x = as.data.frame(results), 
            file = "./counts/results_gene_annotated.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant annotated results table to a .txt file
write.table(x = as.data.frame(results_sig), 
            file = "./counts/results_gene_annotated_significant.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)
```

## Step 9. Plotting Gene Expression Data

### 9a. PCA plot

```{r fig.width=8,fig.height=6}
# Convert all samples to rlog
ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Plot PCA by column variable
plotPCA(ddsMat_rlog, intgroup = "Group", ntop = 500) +
  theme_bw() + # remove default ggplot2 theme
  geom_point(size = 5) + # Increase point size
  #scale_y_continuous(limits = c(-20, 20)) + # change limits to fix figure dimensions
  ggtitle(label = "Principal Component Analysis (PCA)", 
          subtitle = "Top 500 most variable genes") 
```

### 9b. Heatmap

```{r fig.width=8,fig.height=28}
# Convert all samples to rlog
#ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Gather 40 significant genes (by fdr) and make matrix
de.genes <- rownames(results)[which(results$padj < 0.05 & abs(results$log2FoldChange) > 2)]
mat <- assay(ddsMat_rlog[de.genes, ])

# Choose which column variables you want to annotate the columns by.
annotation_col = data.frame(
  Group = factor(colData(ddsMat_rlog)$Group), 
  Replicate = factor(colData(ddsMat_rlog)$Replicate),
  row.names = colData(ddsMat_rlog)$sampleid
)

# Specify colors you want to annotate the columns by.
Group = c(control = "lightblue", case = "darkorange")
names(Group) <- levels(ddsMat$Group)
ann_colors = list(Group,Replicate = c(Rep1 = "darkred", Rep2 = "forestgreen",Rep3 = "darkblue"))

# Make Heatmap with pheatmap function.
## See more in documentation for customization
pheatmap(mat = mat, 
         #color = colorRampPalette(brewer.pal(9, "YlOrBr"))(255), 
         scale = "row", # Scale genes to Z-score (how many standard deviations)
         annotation_col = annotation_col, # Add multiple annotations to the samples
         annotation_colors = ann_colors,# Change the default colors of the annotations
         fontsize = 6.5, # Make fonts smaller
         cellwidth = 55, # Make the cells wider
         show_colnames = T,
         show_rownames = T)
```

### 9c. Volcano Plot

```{r fig.width=8,fig.height=6}
# Gather Log-fold change and FDR-corrected pvalues from DESeq2 results
## - Change pvalues to -log10 (1.3 = 0.05)
data <- data.frame(gene = row.names(results),
                   pval = -log10(results$padj), 
                   lfc = results$log2FoldChange)

# Remove any rows that have NA as an entry
data <- na.omit(data)

# Color the points which are up or down
## If fold-change > 2 and pvalue > 1.3 (Increased significant)
## If fold-change < -2 and pvalue > 1.3 (Decreased significant)
data <- mutate(data, color = case_when(data$lfc > 2 & data$pval > 1.3 ~ "Increased (log2(FC) > 2)",
                                       data$lfc < -2 & data$pval > 1.3 ~ "Decreased (log2(FC) < -2)",
                                       data$lfc >= -2 |  data$lfc <= 2 | data$pval < 1.3 ~ "nonsignificant"))

# Make a basic ggplot2 object with x-y values
vol <- ggplot(data, aes(x = lfc, y = pval, color = color))

# Add ggplot2 layers
vol +   
  ggtitle(label = "Volcano Plot", subtitle = "Colored by fold-change direction") +
  geom_point(size = 1.5, alpha = 0.8, na.rm = T) +
  scale_color_manual(name = "Directionality",
                     values = c("Increased (log2(FC) > 2)"= "#008B00", "Decreased (log2(FC) < -2)" = "#CD4F39", "nonsignificant" = "darkgray")) +
  theme_bw(base_size = 14) + # change overall theme
  theme(legend.position = "right") + # change the legend
  xlab(expression(log[2](case / control))) + # Change X-Axis label
  ylab(expression(-log[10]("adjusted p-value"))) + # Change Y-Axis label
  geom_hline(yintercept = 1.3, colour = "darkgrey") + # Add p-adj value cutoff line
  geom_vline(xintercept = c(-2,2), colour = "darkgrey") + # Add p-adj value cutoff line
  xlim(c(-10,10)) +
  scale_y_continuous(trans = "log1p") # Scale yaxis due to large p-values
```

### 9d. MA Plot

> https://en.wikipedia.org/wiki/MA_plot

```{r fig.width=6,fig.height=6}
# MA Plot
plotMA(results, ylim = c(-10, 10))
```

### 9e. Plot Dispersions

```{r fig.width=6,fig.height=6}
# Plot Dispersions
plotDispEsts(ddsMat)
```

### 9f. Single gene plot

```{r fig.width=6,fig.height=6}
# Single gene plot
# Convert all samples to rlog
#ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Get gene with highest expression
top_gene <- rownames(results)[which.max(results$log2FoldChange)]

# Plot single gene
plotCounts(dds = ddsMat, 
           gene = top_gene, 
           intgroup = "Group", 
           normalized = T, 
           transform = T)
```

## Step 10. Finding Pathways from Differential Expressed Genes

### 10a. Set up matrix to take into account EntrezID's and fold changes for each gene

```{r}
# Remove any genes that do not have any entrez identifiers
results_sig_entrez <- subset(results_sig, is.na(entrez) == FALSE)

# Create a matrix of gene log2 fold changes
gene_matrix <- results_sig_entrez$log2FoldChange

# Add the entrezID's as names for each logFC entry
names(gene_matrix) <- results_sig_entrez$entrez

# View the format of the gene matrix
##- Names = ENTREZ ID
##- Values = Log2 Fold changes
head(gene_matrix)

# choose upregulated genes or downregulated genes!
#gene_matrix <- gene_matrix[gene_matrix > 0]
#gene_matrix <- gene_matrix[gene_matrix < 0]
```
### 10b. Enrich genes using the Gene Onotlogy

```{r fig.width=10,fig.height=6}
#gene_matrix <- gene_matrix[abs(gene_matrix) > 2] # filter DE genesï¼ŒAttention!

if ( !dir.exists("GO") ) {
  dir.create("./GO")
}

for (genegroup in c("total","upregulated","downregulated")) {
  if (genegroup == "total") {
    gene_matrix_filtered <- gene_matrix
  }else if (genegroup == "upregulated") {
    gene_matrix_filtered <- gene_matrix[gene_matrix > 0]
  }else if (genegroup == "downregulated") {
    gene_matrix_filtered <- gene_matrix[gene_matrix < 0]
  }
  
  go_enrich <- enrichGO(gene = names(gene_matrix_filtered),
                      OrgDb = org.db, 
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.10)
  # Plot results
  if (!is.null(go_enrich) & (nrow(go_enrich) != 0)) {
      print(barplot(go_enrich, 
          drop = TRUE, 
          showCategory = 15, 
          title = paste("GO Biological Process using",genegroup,"Genes",sep = " "),
          font.size = 8))
  
      write.csv(go_enrich,file = paste("./GO/GO-Biological-Process-using",genegroup,"Genes.csv",sep = "-"))
  }
}
```

### 10c. Enrich genes using the KEGG database

```{r}
if ( !dir.exists("KEGG") ) {
  dir.create("./KEGG")
}

for (genegroup in c("total","upregulated","downregulated")) {
  if (genegroup == "total") {
    gene_matrix_filtered <- gene_matrix
  }else if (genegroup == "upregulated") {
    gene_matrix_filtered <- gene_matrix[gene_matrix > 0]
  }else if (genegroup == "downregulated") {
    gene_matrix_filtered <- gene_matrix[gene_matrix < 0]
  }

  kegg_enrich <- enrichKEGG(gene = names(gene_matrix_filtered),
                            keyType = "ncbi-geneid",
                            organism = kegg.organism,
                            pvalueCutoff = 0.05,
                            qvalueCutoff = 0.10)

  # Plot results
  if (!is.null(kegg_enrich) & (nrow(kegg_enrich) != 0)) {
    print(barplot(kegg_enrich,
          drop = TRUE,
          showCategory = 15,
          title =  paste("KEGG Enrichment Pathways",genegroup,"Genes",sep = " "),
          font.size = 8))
  write.csv(kegg_enrich,file = paste("./KEGG/KEGG-Enrichment-Pathways",genegroup,"Genes.csv",sep = "-"))

    # Plotting KEGG Pathways
  for (path_id in kegg_enrich@result$ID[kegg_enrich@result$p.adjust < 0.05]) {
    tryCatch({
      suppressMessages(pathview(gene.data = gene_matrix_filtered,
                              pathway.id = path_id,
                              species = kegg.organism))},error = function(e){cat("Error:", conditionMessage(e), '\n')})
    }
  }
}

pathview.files <- grep("pathview.png",list.files(),value = TRUE)
if (length(pathview.files) != 0) {
  for (pathview.file in pathview.files) {
  files.move <- grep(gsub("pathview.png","",pathview.file),list.files(),value = TRUE)
  file.move(files.move,"./KEGG",overwrite = TRUE)
  }
}
```


## step 11. GSEA analysis

```{r}
if ( !dir.exists("GSEA") ) {
  dir.create("./GSEA")
}

# export rnk list using ENSEMBL gene ids
geneList = results[,"log2FoldChange"]
names(geneList) <- rownames(results)
geneList = sort(geneList, decreasing = TRUE)
write.table(geneList,file = "./GSEA/RNAseq-GeneNameENSEMBL-RankedBy-log2FoldChange.rnk",quote = FALSE,col.names = FALSE,sep = "\t")
```

```{r}
# # export rnk list using ENTREZID(NCBI Gene id) and used for downstream analysis
geneList = results[,"log2FoldChange"]
names(geneList) <- results$entrez
geneList <- geneList[!is.na(names(geneList))] # remove genes without NCBI GENE ID, Attention!
head(geneList);length(geneList)
geneList = sort(geneList, decreasing = TRUE)
write.table(geneList,file = "./GSEA/RNAseq-GeneNameENTREZID-RankedBy-log2FoldChange.rnk",quote = FALSE,col.names = FALSE,sep = "\t")
```

### 11.1 Using ClusterProfiler

#### 11.1.1 GO

```{r}
ego2 <- gseGO(geneList     = geneList,
              OrgDb        = org.db,
              keyType      = "ENTREZID",
              ont          = "CC",
              # nPerm        = 1000,
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

ego2@result[,2:7]
```

```{r fig.width=8,fig.height=5}
if (dim(ego2@result)[1] != 0) {
  for (i in 1:dim(ego2@result)[1]) {
    print(enrichplot::gseaplot2(ego2, geneSetID = i,pvalue_table = TRUE))
  }
  write.csv(ego2@result,file = "./GSEA/gseGO-CC.csv")
}
```


```{r}
ego3 <- gseGO(geneList     = geneList,
              OrgDb        = org.db,
              keyType      = "ENTREZID",
              ont          = "BP",
              # nPerm        = 1000,
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

ego3@result[,2:7]
```


```{r}
if (dim(ego3@result)[1] != 0) {
  for (i in 1:dim(ego3@result)[1]) {
    print(enrichplot::gseaplot2(ego3, geneSetID = i,pvalue_table = TRUE))
    write.csv(ego3@result,file = "./GSEA/gseGO-BP.csv")
  }
}
```


#### 11.1.2 KEGG

```{r}
kk2 <- gseKEGG(geneList     = geneList,
               organism     = kegg.organism,
               keyType      = "ncbi-geneid",
               nPerm        = 1000,
               minGSSize    = 50,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
head(kk2)
```

```{r}
if (dim(kk2@result)[1] != 0) {
  for (i in 1:dim(kk2@result)[1]) {
    print(enrichplot::gseaplot2(kk2, geneSetID = i,pvalue_table = TRUE))
    write.csv(kk2@result,file = "./GSEA/gseKEGG.csv")
  }
}
```

#### 11.1.3  MSigDb analysis Only for Human Genes! 

for other species, You should prepare your own gene sets!

The MSigDB is a collection of annotated gene sets, it include 8 major collections:

download from: http://software.broadinstitute.org/gsea/msigdb/index.jsp only for Human genes!

- H: hallmark gene sets
- C1: positional gene sets
- C2: curated gene sets
- C3: motif gene sets
- C4: computational gene sets
- C5: GO gene sets
- C6: oncogenic signatures
- C7: immunologic signatures

```{r}
if (species == "human") {
  # enricher
  gene <- names(geneList)[abs(geneList) > 2]
  c5 <- read.gmt("your-gmtfile.gmt") 
  
  egmt <- enricher(gene, TERM2GENE=c5)
  head(egmt)
  
  if (dim(egmt@result)[1] != 0) {
    print(barplot(egmt, 
        drop = TRUE, 
        showCategory = 15, 
        title = "MSigDb Gene Sets",
        #title = "Using upregulated genes",
        #title = "Using downregulated genes",
        font.size = 8))
    write.csv(egmt@result,file = "enricher-MSigDb.csv")
  }
  
  # GSEA
  egmt2 <- GSEA(geneList, TERM2GENE=c5, verbose=FALSE)
  head(egmt2)
  
  if (dim(egmt2@result)[1] != 0) {
    print(enrichplot::gseaplot2(egmt2, geneSetID = 1,pvalue_table = TRUE))
    write.csv(egmt2@result,file = "GSEA-MSigDb.csv")
  }
}
```


#### 11.1.4  Using your own gene sets

```{r}
# make your own customed gmt file(can be from ATACseq/oDamID...)
if (FALSE) {
  odam <- read.csv("results.annotated.csv",stringsAsFactors = FALSE)
  odam <- filter(odam,log2FoldChange < 0, log2FoldChange > -5)
  #odam <- arrange(odam,desc(log2FoldChange))[1:1000,] # you can use top N, or all.
  #odam <- arrange(odam,desc(score))[1:1000,] # you can use top N, or all.
  
  #hist(odam$log2FoldChange,breaks = 400,xlim = c(-15,5))
  #hist(odam$score,breaks = 100)
  
  write.table(t(c("oDAM-Pros-peaks", # pathway name
                  "peaks from oDAM", # description
                  unique(odam$Gene.Name[odam$Gene.Name%in% names(geneList)]))), # all related genes
                  file = "oDam-pros-log2FoldChange-morethan1.gmt",
                  col.names = FALSE,quote = FALSE,sep = "\t",row.names = FALSE)
}
```

```{r}
# GSEA plot
if (FALSE) {
  c5 <- read.gmt("oDam-pros-log2FoldChange-morethan1.gmt")
  egmt2 <- GSEA(geneList, TERM2GENE=c5, pvalueCutoff = 0.05,maxGSSize = 2300)
  print(enrichplot::gseaplot2(egmt2, geneSetID = 1,pvalue_table = TRUE))
}
```

### 11.2 Using GSEA software in windows

#### 11.2.1 From counts

export count data from DESeq2

```{r}
write.table(x = counts(ddsMat, normalized = T), 
            file = './GSEA/normalized_counts_for_GSEA.txt', 
            sep = '\t', 
            quote = F, 
            col.names = NA)
```

Modify the txt file, adding the first column name!

and make the cls file, refer to:

> http://software.broadinstitute.org/gsea/doc/GSEAUserGuideFrame.html

> http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GCT:_Gene_Cluster_Text_file_format_.28.2A.gct.29

#### 11.2.1 From ranked list

use .rnk files for GSEA analysis!
