---
title: "Step1-Seurat"
author: "Zhang Yongchao"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Seurat - Guided Clustering Tutorial

see: https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html

```{r}
.libPaths(.libPaths()[2])
suppressMessages(library(dplyr)) 
suppressMessages(library(Seurat))
suppressMessages(library(patchwork))
suppressMessages(library(ggplot2))
suppressMessages(library(Matrix))
suppressMessages(library(clustree))
suppressMessages(library(scater))
suppressMessages(library(DT))
```

## import data

```{r}
project.name <- basename(getwd())
matrix.dir <- readRDS("CellRanger_Matrix_path.rds")
matrix.dir
cellranger_mat <- Read10X(data.dir = matrix.dir)
pbmc <- CreateSeuratObject(counts = cellranger_mat, project = project.name)
pbmc
```

## QC

```{r fig.width=8,fig.height=4}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^mt:")
pbmc[["log10GenesPerUMI"]] <- log10(pbmc$nFeature_RNA) / log10(pbmc$nCount_RNA)
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "log10GenesPerUMI"), ncol = 4,pt.size = 0) &   theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

> [hbctraining - 04_SC_quality_control](https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md)

**number of genes detected per UMI**: this metric with give us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)


```{r fig.width=12,fig.height=6}
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
print(plot1 + plot2)
```


```{r fig.width=6,fig.height=6}
plotHighestExprs(as.SingleCellExperiment(pbmc))
```

## save a demo QC matrix

```{r}
source("./functions.R")
pbmc <- qc_filter(obj = pbmc,
                  nFeature_RNA.min = 800,
                  nFeature_RNA.max = 5000,
                  nCount_RNA.min = 3000,
                  nCount_RNA.max = 2e4,
                  percent.mt.max = 15,
                  log10GenesPerUMI.min = 0.75)
```

```{r fig.width=8,fig.height=4}
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "log10GenesPerUMI"), ncol = 4,pt.size = 0) &   theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```
```{r fig.width=12,fig.height=6}
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
print(plot1 + plot2)
```

## Normalize data

```{r}
pbmc <- NormalizeData(pbmc)
```

## Identification of highly variable features (feature selection)

```{r fig.width=10,fig.height=5}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top20 <- head(VariableFeatures(pbmc), 20)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc) + NoLegend()
plot2 <- LabelPoints(plot = plot1, points = top20, repel = TRUE) + NoLegend()
plot1 + plot2
```
## Scaling the data

```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

## Perform linear dimensional reduction

```{r fig.width=10,fig.height=8}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
```

```{r fig.width=7,fig.height=7}
DimPlot(pbmc, reduction = "pca") + NoLegend()
```
```{r fig.width=7,fig.height=7}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```
```{r fig.width=12,fig.height=20}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```
## Determine the ¡®dimensionality¡¯ of the dataset

```{r fig.height=5,fig.width=7}
ElbowPlot(pbmc)
```
## Cluster the cells

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:20) %>% FindClusters(resolution = c(0.2,0.4,0.6,0.8,1.0,1.2,1.4,1.6,1.8,2))
```
## Run non-linear dimensional reduction (UMAP/tSNE)

```{r fig.width=6,fig.height=5}
Idents(pbmc) <- "RNA_snn_res.1"
pbmc <- RunUMAP(pbmc, dims = 1:20)
pbmc <- RunTSNE(pbmc, dims = 1:20)
DimPlot(pbmc, reduction = "umap")
DimPlot(pbmc, reduction = "tsne")
```

## Finding differentially expressed features (cluster biomarkers)

```{r}
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %>%
  dplyr::select(gene,cluster,everything()) %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1) %>% 
  ungroup() %>%
  datatable(rownames = FALSE) %>% 
  formatPercentage("pct.1",2) %>% 
  formatPercentage("pct.2",2) %>% 
  formatSignif("p_val", digits = 2) %>%
  formatSignif("p_val_adj", digits = 2) %>%
  formatRound('avg_log2FC', 3)
```

```{r fig.width=20,fig.height=15}
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```



## save rds and markers

```{r}
# save rds
rds.dir <- "./1.Rds-file/"
if ( !dir.exists(rds.dir) ) {
 dir.create(rds.dir)
}
saveRDS(pbmc, file = paste0(rds.dir,"/",project.name,".rds"))
# save marker
marker.dir <- "./2.Cluster-markers/"
if ( !dir.exists(marker.dir) ) {
 dir.create(marker.dir)
}
write.csv(pbmc.markers, file = paste0(marker.dir,"/cluster-markers-of-resolution-1.0.csv"))
```

