---
title: "ATACseqQC Guide"
author: "Zhang Yongchao"
date: "r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Assessing other QC metrics**

> [Best practices on ATAC-seq QC and data analysis](https://haibol2016.github.io/ATACseqQCWorkshop/articles/ATACseqQC_workshop.html)

A few new QC functions have been added to the ATACseqQC package.

- distanceDyad: calculate the distance of potential nucleosome dyad and the linear model for V.

- NFRscore: calculate the ratio of cutting signal immediately adjacent to TSSs and that located to the regions flanking TSSs.

- PTscore: calculate the ratio of read coverage over promoters to read coverage over transcript body.

- TSSEscore: calculate aggregated distribution of reads centered on TSSs and that of reads flanking the corresponding TSSs.

- vPlot: aggregate ATAC-seq Fragment Midpoint vs. Length for a given motif generated over binding sites within the genome.

> [ATACseqQC Guide](https://bioconductor.org/packages/release/bioc/vignettes/ATACseqQC/inst/doc/ATACseqQC.html)
> ATACseqQC Guide
> Jianhong Ou, Haibo Liu, Jun Yu, Michelle Kelliher, Lucio Castilla, Nathan Lawson, Lihua Julie Zhu 26 October 2021

# Introduction

Assay for Transposase-Accessible Chromatin using sequencing (ATAC-seq) is an alternative or complementary technique to MNase-seq, DNase-seq, and FAIRE-seq for chromatin accessibility analysis. The results obtained from ATAC-seq are similar to those from DNase-seq and FAIRE-seq. ATAC-seq is gaining popularity because it does not require cross-linking, has higher signal to noise ratio, requires a much smaller amount of biological material and is faster and easier to perform, compared to other techniques1.

To help researchers quickly assess the quality of ATAC-seq data, we have developed the ATACseqQC package for easily making diagnostic plots following the published guidelines1. In addition, it has functions to preprocess ATACseq data for subsequent peak calling.

# Quick start

Here is an example using ATACseqQC with a subset of published ATAC-seq data1. Currently, only bam input file format is supported.

First install ATACseqQC and other packages required to run the examples. Please note that the example dataset used here is from human. To run analysis with dataset from a different species or differnt assembly, please install the corresponding BSgenome, TxDb and phastCons. For example, to analyze mouse data aligned to mm10, please install BSgenome.Mmusculus.UCSC.mm10, TxDb.Mmusculus.UCSC.mm10.knownGene and phastCons60way.UCSC.mm10. Please note that phstCons60way.UCSC.mm10 is optional, which can be obtained according to the vignettes of GenomicScores.

```{r}
# library(BiocManager)
# BiocManager::install(c("ATACseqQC", "ChIPpeakAnno", "MotifDb", "GenomicAlignments",
#            "BSgenome.Hsapiens.UCSC.hg19", "TxDb.Dmelanogaster.UCSC.dm6.ensGene",
#            "phastCons100way.UCSC.hg19"))
```

```{r}
## load the library
library(ATACseqQC)
## input the bamFile from the ATACseqQC package 
bamfiles <- list.files("../3_aligned_Bowtie2/raw_bam/",pattern = "bam$",full.names = TRUE)[1]
bamfiles
bamfiles.labels <- gsub(".sorted.bam", "", basename(bamfiles))
```


## IGV snapshot

Source code of IGVSnapshot function is available in extdata folder. To call the function, please try

```{r}
# source(system.file("extdata", "IGVSnapshot.R", package = "ATACseqQC"))
```


## Estimate the library complexity

whether the sequencing depth is saturated or not, nor whether the library is complex enough for deeper sequencing.

CAUTION: only BAM files without removing duplicates are informative for estimating library complexity.

```{r}
sapply(bamfiles, function(x)bamQC(x, outPath=NULL)) # too slow, and not a necessary step!
bamfiles
sapply(bamfiles, function(x)estimateLibComplexity(readsDupFreq(x))) # add sample name to the plot
```

## Fragment size distribution

First, there should be a large proportion of reads with less than 100 bp, which represents the nucleosome-free region. Second, the fragment size distribution should have a clear periodicity, which is evident in the inset figure, indicative of nucleosome occupacy (present in integer multiples).

Please note the **pre-filtered BAM files** need to be used to get an unbiased distribution of insert fragment size in the ATAC-seq library.

```{r}
## generate fragment size distribution
bamfiles <- list.files("../3_aligned_Bowtie2",pattern = "rmBlacklist.bam$",full.names = TRUE)
bamfiles.labels <- gsub(".rmChrM.rmDup.rmMulti.sorted.rmBlacklist.bam", "", basename(bamfiles))
fragSize <- fragSizeDist(bamFiles = bamfiles, bamfiles.labels)
```

## Nucleosome positioning

### Adjust the read start sites

Tn5 transposase has been shown to bind as a dimer and inserts two adaptors into accessible DNA locations separated by 9 bp2.

Therefore, for downstream analysis, such as peak-calling and footprinting, all reads in input bamfile need to be shifted. The function shiftGAlignmentsList can be used to shift the reads. By default, all reads aligning to the positive strand are offset by +4bp, and all reads aligning to the negative strand are offset by -5bp1.

The adjusted reads will be written into a new bamfile for peak calling or footprinting.

```{r}
## bamfile tags to be read in
possibleTag <- list("integer"=c("AM", "AS", "CM", "CP", "FI", "H0", "H1", "H2", 
                                "HI", "IH", "MQ", "NH", "NM", "OP", "PQ", "SM",
                                "TC", "UQ"), 
                 "character"=c("BC", "BQ", "BZ", "CB", "CC", "CO", "CQ", "CR",
                               "CS", "CT", "CY", "E2", "FS", "LB", "MC", "MD",
                               "MI", "OA", "OC", "OQ", "OX", "PG", "PT", "PU",
                               "Q2", "QT", "QX", "R2", "RG", "RX", "SA", "TS",
                               "U2"))
library(Rsamtools)
bamTop100 <-lapply(bamfiles, function(x){
  scanBam(BamFile(x, yieldSize = 100),
                     param = ScanBamParam(tag=unlist(possibleTag)))[[1]]$tag
})
tags <- lapply(bamTop100, function(x){
  names(x)[lengths(x)>0]
})
names(tags) <- bamfiles.labels
```

```{r}
## files will be output into outPath
outPath <- "ATACseqQC-shifted"
dir.create(outPath)
## shift the coordinates of 5'ends of alignments in the bam file
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
seqlev <- "chr2L" ## subsample data for quick run
seqinformation <- seqinfo(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
which <- as(seqinformation[seqlev], "GRanges")
## if you don't have an available TxDb, please refer
## GenomicFeatures::makeTxDbFromGFF to create one from gff3 or gtf file.
shift_bam <- function(x, tags = "tags", outPath = "outPath", outbam.name = "shifted.bam"){
  # gal <- readBamFile(x, tag=tags, asMates=TRUE, bigFile=TRUE)
  gal <- readBamFile(x, tag=tags, asMates=TRUE, bigFile=TRUE, which = which)
  shiftedBamfile <- file.path(outPath, outbam.name)
  shiftGAlignmentsList(gal, outbam=shiftedBamfile)
}
library(parallel)
gal.list <- mclapply(1:length(bamfiles), function(x){
  shift_bam(bamfiles[x], tags = tags[[x]], outPath = outPath,outbam.name = paste0(bamfiles.labels[x],".bam"))
  },mc.cores = 12)
```


### Promoter/Transcript body (PT) score

PT score is calculated as the coverage of promoter divided by the coverage of its transcript body. PT score will show if the signal is enriched in promoters.


```{r}
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
txs <- transcripts(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
mclapply(gal.list, function(x){
  pt <- PTscore(x, txs) # long time
  print(plot(pt$log2meanCoverage, pt$PT_score, 
     xlab="log2 mean coverage",
     ylab="Promoter vs Transcript",
     main = gsub(".bam","",basename(x@metadata$file))))
},mc.cores = 12)
```


### Nucleosome Free Regions (NFR) score

NFR score is a ratio between cut signal adjacent to TSS and that flanking the corresponding TSS. Each TSS window of 400 bp is first divided into 3 sub-regions: the most upstream 150 bp (n1), the most downstream of 150 bp (n2), and the middle 100 bp (nf). Then the number of fragments with 5’ ends overlapping each region are calculated for each TSS. The NFR score for each TSS is calculated as NFR-score = log2(nf) - log2((n1+n2)/2). A plot can be generated with the NFR scores as Y-axis and the average signals of 400 bp window as X-axis, very like a MA plot for gene expression data.

```{r}
mclapply(gal.list,function(x){
  nfr <- NFRscore(x, txs) # long time!
  print(plot(nfr$log2meanCoverage, nfr$NFR_score, 
     xlab="log2 mean coverage",
     ylab="Nucleosome Free Regions score",
     main=paste0("NFRscore for 200bp flanking TSSs of ",gsub(".bam","",basename(x@metadata$file))),
     xlim=c(-10, 0), ylim=c(-5, 5)))
},mc.cores = 12)

```

### Transcription Start Site (TSS) Enrichment Score

TSS enrichment score is a raio between aggregate distribution of reads centered on TSSs and that flanking the corresponding TSSs. TSS score = the depth of TSS (each 100bp window within 1000 bp each side) / the depth of end flanks (100bp each end). TSSE score = max(mean(TSS score in each window)). TSS enrichment score is calculated according to the definition at https://www.encodeproject.org/data-standards/terms/#enrichment. Transcription start site (TSS) enrichment values are dependent on the reference files used; cutoff values for high quality data are listed in the following table from https://www.encodeproject.org/atac-seq/.

```{r}
mclapply(gal.list,function(x){
  tsse <- TSSEscore(x, txs)
  print(tsse$TSSEscore)
  print(plot(100*(-9:10-.5), tsse$values, type="b",
     xlab="distance to TSS",
     ylab="aggregate TSS score",
     main = gsub(".bam","",basename(x@metadata$file))))
},mc.cores = 12)
```


### Split reads

The shifted reads will be split into different bins, namely nucleosome free, mononucleosome, dinucleosome, and trinucleosome. Shifted reads that do not fit into any of the above bins will be discarded. Splitting reads is a time-consuming step because we are using random forest to classify the fragments based on fragment length, GC content and conservation scores3.

By default, we assign the top 10% of short reads (reads below 100_bp) as nucleosome-free regions and the top 10% of intermediate length reads as (reads between 180 and 247 bp) mononucleosome. This serves as the training set to classify the rest of the fragments using random forest. The number of the tree will be set to 2 times of square root of the length of the training set.

```{r}
library(BSgenome.Dmelanogaster.UCSC.dm6)
# library(phastCons100way.UCSC.hg19)
## run program for chromosome 1 only
# txs <- txs[seqnames(txs) %in% "chr1"]
# genome <- Hsapiens
# ## split the reads into NucleosomeFree, mononucleosome, 
# ## dinucleosome and trinucleosome.
# ## and save the binned alignments into bam files.
# objs <- splitGAlignmentsByCut(gal1, txs=txs, genome=genome, outPath = outPath,
#                               conservation=phastCons100way.UCSC.hg19)
# ## list the files generated by splitGAlignmentsByCut.
# dir(outPath)
```

You can also perform shifting, splitting and saving in one step by calling splitBam.

```{r}
# objs <- splitBam(bamfile, tags=tags, outPath=outPath,
#                  txs=txs, genome=genome,
#                  conservation=phastCons100way.UCSC.hg19)
```

Conservation is an optional parameter. If you do not have the conservation score or you would like to simply split the bam files using the fragment length, then you will just need to run the command without providing the conservation argument. Without setting the conservation parameter, it will run much faster.

```{r}
## split reads by fragment length
## NOT RUN IN THIS example
# objs.list <- mclapply(gal.list,function(x){
#   save.dir <- paste0("./ATACseqQC-shifted/",gsub(".bam","",basename(x@metadata$file)))
#   dir.create(save.dir, recursive = TRUE )
#   splitGAlignmentsByCut(x, txs=txs, outPath = save.dir)
# }, mc.cores = 12)
```


### Heatmap and coverage curve for nucleosome positions

By averaging the signal across all active TSSs, we should observe that nucleosome-free fragments are enriched at the TSSs, whereas the nucleosome-bound fragments should be enriched both upstream and downstream of the active TSSs and display characteristic phasing of upstream and downstream nucleosomes. Because ATAC-seq reads are concentrated at regions of open chromatin, users should see a strong nucleosome signal at the +1 nucleosome, but the signal decreases at the +2, +3 and +4 nucleosomes.


>>>>>>>>>>>>>>>>> 代码修改到这里了 - 2022.1.11 放弃 <<<<<<<<<<<<<<<<<<

```{r}
library(ChIPpeakAnno)
TSS <- promoters(txs, upstream=0, downstream=1)
TSS <- unique(TSS)
  
mclapply(gal.list,function(x){
  save.dir <- paste0("./ATACseqQC-shifted/",gsub(".bam","",basename(x@metadata$file)))
  dir.create(save.dir, recursive = TRUE )
  objs <- splitGAlignmentsByCut(x, txs=txs, outPath = save.dir)
  bamfiles <- file.path(save.dir,
                     c("NucleosomeFree.bam",
                     "mononucleosome.bam",
                     "dinucleosome.bam",
                     "trinucleosome.bam"))
  ## Plot the cumulative percentage of tag allocation in nucleosome-free 
  ## and mononucleosome bam files.
  print(cumulativePercentage(bamfiles[1:2], as(seqinformation["chr2L"], "GRanges")))
 
  ## estimate the library size for normalization
  librarySize <- estLibSize(bamfiles)
  ## calculate the signals around TSSs.
  NTILE <- 101
  dws <- ups <- 1010
  sigs <- enrichedFragments(gal=objs[c("NucleosomeFree", 
                                       "mononucleosome",
                                       "dinucleosome",
                                       "trinucleosome")], 
                            TSS=TSS,
                            librarySize=librarySize,
                            seqlev=seqlev,
                            TSS.filter=0.5,
                            n.tile = NTILE,
                            upstream = ups,
                            downstream = dws)
  ## log2 transformed signals
  sigs.log2 <- lapply(sigs, function(.ele) log2(.ele+1))
  #plot heatmap
  print(featureAlignedHeatmap(sigs.log2, reCenterPeaks(TSS, width=ups+dws),
                        zeroAt=.5, n.tile=NTILE))
  
  ## get signals normalized for nucleosome-free and nucleosome-bound regions.
  out <- featureAlignedDistribution(sigs, 
                                    reCenterPeaks(TSS, width=ups+dws),
                                    zeroAt=.5, n.tile=NTILE, type="l", 
                                    ylab="Averaged coverage")
  ## rescale the nucleosome-free and nucleosome signals to 0~1
  range01 <- function(x){(x-min(x))/(max(x)-min(x))}
  out <- apply(out, 2, range01)
  print(matplot(out, type="l", xaxt="n", 
          xlab="Position (bp)", 
          ylab="Fraction of signal"))
  print(axis(1, at=seq(0, 100, by=10)+1, 
       labels=c("-1K", seq(-800, 800, by=200), "1K"), las=2))
  print(abline(v=seq(0, 100, by=10)+1, lty=2, col="gray"))
  
}, mc.cores = 12)

```




## plot Footprints

ATAC-seq footprints infer factor occupancy genome-wide. The factorFootprints function uses matchPWM to predict the binding sites using the input position weight matrix (PWM). Then it calculates and plots the accumulated coverage for those binding sites to show the status of the occupancy genome-wide. Unlike CENTIPEDE4, the footprints generated here do not take the conservation (PhyloP) into consideration. factorFootprints function could also accept the binding sites as a GRanges object.

```{r}
## foot prints
library(MotifDb)
CTCF <- query(MotifDb, c("CTCF"))
CTCF <- as.list(CTCF)
print(CTCF[[1]], digits=2)
```

```{r}
sigs <- factorFootprints(shiftedBamfile, pfm=CTCF[[1]], 
                         genome=genome, ## Don't have a genome? ask ?factorFootprints for help
                         min.score="90%", seqlev=seqlev,
                         upstream=100, downstream=100)
```

```{r}
featureAlignedHeatmap(sigs$signal, 
                      feature.gr=reCenterPeaks(sigs$bindingSites,
                                               width=200+width(sigs$bindingSites[1])), 
                      annoMcols="score",
                      sortBy="score",
                      n.tile=ncol(sigs$signal[[1]]))
```

```{r}
sigs$spearman.correlation
```

```{r}
sigs$Profile.segmentation
```


Here is the CTCF footprints for the full dataset. CTCF footprints

### V-plot

V-plot is a plot to visualize fragment midpoint vs length for a given transcription factors.

```{r}
vp <- vPlot(shiftedBamfile, pfm=CTCF[[1]], 
            genome=genome, min.score="90%", seqlev=seqlev,
            upstream=200, downstream=200, 
            ylim=c(30, 250), bandwidth=c(2, 1))
```

```{r}
distanceDyad(vp, pch=20, cex=.5)
```

Here is the CTCF vPlot for the full dataset. CTCF footprints

# Plot correlations for multiple samples

```{r}
path <- system.file("extdata", package="ATACseqQC", mustWork=TRUE)
bamfiles <- dir(path, "*.bam$", full.name=TRUE)
gals <- lapply(bamfiles, function(bamfile){
               readBamFile(bamFile=bamfile, tag=character(0), 
                          which=GRanges("chr1", IRanges(1, 1e6)), 
                          asMates=FALSE)
         })
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
txs <- transcripts(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(GenomicAlignments)
plotCorrelation(GAlignmentsList(gals), txs, seqlev="chr1")
```


# Session Info

```{r}
sessionInfo()
```

# References

1. Buenrostro, J. D., Giresi, P. G., Zaba, L. C., Chang, H. Y. & Greenleaf, W. J. Transposition of native chromatin for fast and sensitive epigenomic profiling of open chromatin, dna-binding proteins and nucleosome position. Nature methods 10, 12131218 (2013).

2. Adey, A. et al. Rapid, low-input, low-bias construction of shotgun fragment libraries by high-density in vitro transposition. Genome biology 11, R119 (2010).

3. Chen, K. et al. DANPOS: Dynamic analysis of nucleosome position and occupancy by sequencing. Genome research 23, 341351 (2013).

4. Pique-Regi, R. et al. Accurate inference of transcription factor binding from dna sequence and chromatin accessibility data. Genome research 21, 447455 (2011).