---
title: "Step4.2 Downstream Analysis - DEGs"
author: "Zhang Yongchao"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      collapse = TRUE,
                      warning = FALSE)
```

**Parameters:**

```{r}
Results.dir <- "./4.2-Downstream-Analysis/DEGs/Subset-Para"
Rds.dir <- "./4.1-Merged-Rds"
Group.by <- "group"
group.levels <- c("Ctrl","Runx1") # control level in front
Meta.to.regress <- character(0) # Default
#Meta.to.regress <- "batch" # must by numeric
species <- "mouse" # or human, mouse
```


```{r}
suppressMessages(library(Seurat))
suppressMessages(library(dplyr))
suppressMessages(library(patchwork))
suppressMessages(library(cowplot))
suppressMessages(library(ggplot2))
kegg.organism <- switch(species,
                        drosophlia = "dme",
                        human = "hsa",
                        mouse = "mmu")
org.db <- switch(species,
                 drosophlia = "org.Dm.eg.db",
                 human = "org.Hs.eg.db",
                 mouse = "org.Mm.eg.db")
suppressMessages(library(clusterProfiler))
switch (species,
  drosophlia = suppressMessages(library(org.Dm.eg.db)),
  human = suppressMessages(library(org.Hs.eg.db)),
  mouse = suppressMessages(library(org.Mm.eg.db))
)
suppressMessages(library(pathview))
suppressMessages(library(filesstrings))
```

```{r fig.width=6,fig.height=6}
cds <- readRDS(list.files(Rds.dir,pattern = "rds$",full.names = TRUE))
cluster.resolution <- "Merged_Harmony_RNA_snn_res.1"
Idents(cds) <- cluster.resolution
DimPlot(cds,label = TRUE) + NoLegend()
```

```{r fig.height=4,fig.width=8}
DimPlot(cds,label = TRUE,split.by = Group.by) + NoLegend()
```

## Identify differential expressed genes across conditions

### average expression

take the average expression of both the stimulated and control naive T cells and CD14 monocyte populations and generate the scatter plots, highlighting genes that exhibit dramatic responses to interferon stimulation.

```{r fig.height=6,fig.width=6}
theme_set(theme_cowplot())

for (i in 1:length(levels(cds@active.ident))) {
  t.cells <- subset(cds, idents = i-1)
  if (length(unique(t.cells@meta.data[,Group.by])) != 2) {
    next()
  }
  Idents(t.cells) <- "Tmp"
  avg.t.cells <- log1p(AverageExpression(t.cells, verbose = FALSE,add.ident = Group.by,assays = DefaultAssay(cds))[[DefaultAssay(cds)]])
  avg.t.cells$gene <- rownames(avg.t.cells)
  #print(head(avg.t.cells))
  Data.colNames <- unlist(lapply(strsplit(colnames(avg.t.cells)[1:2],split = "_"), "[",2))
  colnames(avg.t.cells)[1:2] <- c("control","case")
  # find genes with significant changes
  test <- avg.t.cells
  test$diff <- abs(test[,1] - test[,2])
  test <- dplyr::arrange(test,desc(diff))
  test <- filter(test,diff >= 1)
  genes.to.label = test$gene
  if(length(genes.to.label >= 30)){genes.to.label <- genes.to.label[1:30]}

  p1 <- ggplot(avg.t.cells, aes(control,case)) + geom_point() + ggtitle(paste("cluster-",i-1,"-By-",Group.by,sep = "")) + xlab(Data.colNames[1]) + ylab(Data.colNames[2])
  if (length(genes.to.label) != 0) {
    p1 <- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE, col = "red")
  }
  print(p1)
}

```

### single cell level for each cluster 

```{r}
cds$celltype <- Idents(cds)
Idents(cds) <- paste(cds@meta.data[,Group.by],Idents(cds), sep = "_")

cds@meta.data[,Group.by] <- factor(as.character(cds@meta.data[,Group.by]), levels = group.levels)

top2Diff.type <- c()
top2Diff.type.regress <- c()

# cds@meta.data$location.regress <- as.numeric(as.factor(cds@meta.data$location))
# cds@meta.data$sex.regress <- as.numeric(as.factor(cds@meta.data$sex))

if ( !dir.exists(Results.dir)) {
  dir.create(Results.dir, recursive = TRUE)
}

for (i in (1:length(levels(cds$celltype)))) {
  samplesSelected <- paste(levels(cds@meta.data[,Group.by]),i-1,sep = "_")
  message(paste("Using",samplesSelected[2],"as Case!",sep = " "))
  if (sum(as.character(Idents(cds)) %in% samplesSelected[1]) < 3 | sum(as.character(Idents(cds)) %in% samplesSelected[2]) < 3) {
    message(paste("Escaped cluster",i-1,"for very few cells!",sep = " "))
    next()
  }
  b.interferon.response <- suppressWarnings(suppressMessages(FindMarkers(cds, ident.1 = samplesSelected[2], ident.2 = samplesSelected[1], verbose = FALSE,test.use = "MAST")) )
  write.csv(b.interferon.response, file = paste(Results.dir,"/DEGs-of-cluster-",i-1,"-",samplesSelected[2],"-vs-",samplesSelected[1],".By.MAST.csv",sep = ""))
  top2Diff.type <- append(top2Diff.type,rownames(b.interferon.response)[1:3])
  
  if (length(Meta.to.regress) !=0 ) {
    b.interferon.response <- suppressWarnings(suppressMessages(FindMarkers(cds, ident.1 = samplesSelected[2], ident.2 = samplesSelected[1], verbose = FALSE,test.use = "MAST",latent.vars = Meta.to.regress)) )
    write.csv(b.interferon.response, file = paste(Results.dir,"/DEGs-of-cluster-",i-1,"-",samplesSelected[2],"-vs-",samplesSelected[1],".By.MAST.",paste(Meta.to.regress,collapse = "."),".Regressd.csv",sep = ""))
    top2Diff.type.regress <- append(top2Diff.type.regress,rownames(b.interferon.response)[1:3])
  }  
}

Idents(cds) <- "celltype"
```

```{r fig.height=6, fig.width=12}
for ( gene in unique(top2Diff.type)) {
  print(FeaturePlot(cds, features = gene, split.by = Group.by, pt.size = 0.5, cols = c("grey", "red")))
}
```

```{r fig.height=6, fig.width=12}
for ( gene in unique(top2Diff.type.regress)) {
  print(FeaturePlot(cds, features = gene, split.by = Group.by, pt.size = 0.5, cols = c("grey", "red")))
}
```

# GO And KEGG analysis

# GO and KEGG Analysis of the DEGs

```{r}
DEGs.files <- list.files(Results.dir,pattern = "^DEGs.*csv$",full.names = TRUE)
```

```{r fig.width=10,fig.height=8}
if ( !dir.exists(paste(Results.dir,"GO",sep = "/")) ) {
  dir.create(paste(Results.dir,"GO",sep = "/"))
}
  
if ( !dir.exists(paste(Results.dir,"KEGG",sep = "/")) ) {
  dir.create(paste(Results.dir,"KEGG",sep = "/"))
}

for (i in DEGs.files) {
  print(paste("Processing: ",i,sep = ""))
  DEGs <- read.csv(i,row.names = 1,stringsAsFactors = FALSE)
  
  # prepare for GO and KEGG
  DEGs$entrez <- mapIds(x = get(org.db),
                     keys = rownames(DEGs),
                     column = "ENTREZID",
                     keytype = "SYMBOL",
                     multiVals = "first")
  # DEGs upregulated and downregulated
  DEGs.up <- DEGs[DEGs$avg_logFC > 0,]
  DEGs.down <- DEGs[DEGs$avg_logFC < 0,]

  # GO - Upregulated
  go_enrich <- enrichGO(gene = DEGs.up$entrez,
                            OrgDb = org.db, 
                            readable = T,
                            ont = "BP",
                            pvalueCutoff = 0.05, 
                            qvalueCutoff = 0.10)
  # Plot results
  if (!is.null(go_enrich)) {
      print(barplot(go_enrich, 
          drop = TRUE, 
          showCategory = 30, 
          title = paste("GO BP Up - ",gsub(".csv$","",gsub("DEGs-of-","",basename(i))),sep = ""),
          font.size = 8))
          write.csv(go_enrich,file = paste(Results.dir,"/GO/GO-BP-Upregulated-of-",basename(i),sep = ""))
    }
  
  # GO - Downregulated
  go_enrich_down <- enrichGO(gene = DEGs.down$entrez,
                            OrgDb = org.db, 
                            readable = T,
                            ont = "BP",
                            pvalueCutoff = 0.05, 
                            qvalueCutoff = 0.10)
  # Plot results
  if (!is.null(go_enrich_down)) {
      print(barplot(go_enrich_down, 
          drop = TRUE, 
          showCategory = 30, 
          title = paste("GO BP Down - ",gsub(".csv$","",gsub("DEGs-of-","",basename(i))),sep = ""),
          font.size = 8))
          write.csv(go_enrich_down,file = paste(Results.dir,"/GO/GO-BP-Downregulated-of-",basename(i),sep = ""))
  }
  
  # KEGG - Up
  kegg_enrich <- enrichKEGG(gene =  DEGs.up$entrez,
                            keyType = "ncbi-geneid",
                            organism = kegg.organism,
                            pvalueCutoff = 0.05, 
                            qvalueCutoff = 0.30)
        
  # Plot results
  if (!is.null(kegg_enrich)) {
    print(barplot(kegg_enrich, 
          drop = TRUE, 
          showCategory = 15, 
          title =  paste("KEGG Up - ",gsub(".csv$","",gsub("DEGs-of-","",basename(i))),sep = ""),
          font.size = 8))
  write.csv(kegg_enrich,file =  paste(Results.dir,"/KEGG/KEGG-Upregulated-",basename(i),".csv",sep = ""))
  }
  
  # KEGG - Down
  kegg_enrich_down <- enrichKEGG(gene =  DEGs.down$entrez,
                            keyType = "ncbi-geneid",
                            organism = kegg.organism,
                            pvalueCutoff = 0.05, 
                            qvalueCutoff = 0.30)
        
  # Plot results
  if (!is.null(kegg_enrich_down)) {
    print(barplot(kegg_enrich_down, 
          drop = TRUE, 
          showCategory = 15, 
          title =  paste("KEGG Up - ",gsub(".csv$","",gsub("DEGs-of-","",basename(i))),sep = ""),
          font.size = 8))
  write.csv(kegg_enrich_down,file =  paste(Results.dir,"/KEGG/KEGG-Downregulated-",basename(i),".csv",sep = ""))
  }
  rm("kegg_enrich","kegg_enrich_down","go_enrich_down","go_enrich")
}
```

# Session info

```{r}
sessionInfo()
```

