---
title: "Step5.2.3 Downstream Analysis - psupertime"
author: "Zhang Yongchao"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      collapse = TRUE,
                      warning = FALSE)
```

>[psupertime](https://github.com/wmacnair/psupertime)

psupertime is an R package which uses single cell RNAseq data, where the cells have labels following a known sequence (e.g. a time series), to identify a small number of genes which place cells in that known order. It can be used for discovery of relevant genes, for exploration of unlabelled data, and assessment of one dataset with respect to the labels known for another dataset.

**Parameters:**

```{r}
# Results.dir <- "./5.2-Downstream-Analysis/5.2.3-psupertime"
```


```{r}
suppressMessages(library(Seurat))
suppressMessages(library(psupertime))
suppressMessages(library(SingleCellExperiment))
```

# import data

```{r fig.width=6,fig.height=6}
cds <- readRDS(list.files("./5.1-Rds",pattern = "rds$",full.names = TRUE))
cluster.resolution <- paste(DefaultAssay(cds),"_snn_res.0.6",sep = "")
Idents(cds) <- cluster.resolution
DimPlot(cds,label = TRUE) + NoLegend()
```

# psupertime

## Model and diagnostics

```{r}
sce <- as.SingleCellExperiment(cds)
day <- cds@meta.data[["day"]]
```

The plot below shows how several measures of performance are affected by the extent of regularization, λ. The x-axis shows λ, indicating how strongly the model tries to set coefficients to zero. The optimal value of λ is the one which gives the best mean performance over the training data, based on one of two possible measures of performance.

```{r}
sce_best = psupertime(sce, day, 
                      sel_genes="hvg",
                      # sel_genes='all', 
                      penalization='best')
plot_train_results(sce_best)
print(sce_best)
```
The first row shows classification error, namely the proportion of cells for which psupertime predicted the wrong label (equivalent to 1 - accuracy). The second row is cross-entropy, which quantifies how confidently the psupertime classifier predicts the correct label (so predicting the correct label with probability p=0.9 results in a lower cross-entropy than with probability p=0.5). Accuracy is a ‘lumpy’ measurement of performance (something is either correct or not), whereas cross-entropy is continuous; this means that selecting λ on the basis of cross-entropy results in less noisy selection of the λ value.

The third row shows the number of genes with non-zero coefficients, for each given value of λ (this is effectively the inverse of sparsity, which is the proportion of zero coefficients).

The solid vertical grey line shows the value of λ resulting in the best performance. The dashed vertical grey line shows the largest value of λ with performance within one standard error of this. By default psupertime selects this value, giving increased sparsity at a minimal cost to performance. We show lines for selection using both classification error and cross-entropy; the thicker lines indicate which measure was actually used to select λ. In this case we used the λ value within 1 s.e. of the best performance on cross-entropy. Reading down to the plot of non-zero genes, we can see that this resulted in just under 100 genes with non-zero coefficients.

## psupertime ordering of cells

```{r fig.width=10,fig.height=6}
plot_labels_over_psupertime(sce_best, label_name='day')
plot_labels_over_psupertime(sce_best, label_name='day',palette="Set1")
```

Interesting things you might observe:

Individual cells may have earlier or later values than others with the same label, possibly suggesting interesting subpopulations within a group label.
The thresholds learned by psupertime indicate how easy it is to distinguish between the different labels: where thresholds are close together, these labels are hard to separate, and where they are distant this task is easier.

## Genes identified by psupertime

psupertime identifies a small set of genes which place the individual cells approximately in the order of the group-level labels. This list can be the most relevant output from psupertime. The plot below shows the 20 genes with the largest absolute coefficient values (subject to the absolute value being >0.05). Genes with positive coefficients will have expression positively correlated with the group-level labels, and vice versa for negative coefficients.

```{r fig.width=10,fig.height=4}
g = plot_identified_gene_coefficients(sce_best, n = 30)
g
```

Another way of examining these genes is to plot their expression values against the learned pseudotime values. The plot below shows the same set of genes, with the (z-scored log) expression values for all individual cells. This can show different profiles of expression, e.g. initially on, then switched off (ITM2A); and increasing or decreasing relatively constantly (CLU).

```{r fig.width=17,fig.height=13}
plot_identified_genes_over_psupertime(sce_best, label_name='day')
plot_identified_genes_over_psupertime(sce_best, label_name='day',palette="Set1")
```

Such gene plots can also potentially identify branching, for example where expression of a given gene is initially unimodal, but later becomes bimodal.

```{r}
plot_specified_genes_over_psupertime(sce_best,
                                     c("Aqp8", "Muc2", "Uba52","Eno1", "A100a11", "Acin1", 
                                       "Ccl20", "Marcksl1", "Chga", "klk1", "Sox11", "Sox4", "Rpl36", "Hnrnpu",
                                       "Hnrnpl", "Lgr4", "Tacstd2", "Chgb", "Neurod1", "Lars2"), label_name ='day',palette="Set1")
```

## psupertime as a classifier

psupertime is a classifier, in the sense that once trained, it can predict a label for any cell given as input. Comparing the predicted classes of cells against their known classes can identify interesting subpopulations of cells.

In the plot below, the x-axis shows the labels used to train psupertime; the y-axis shows the labels of the data used as input for this instance of psupertime (which in this case are the same as the predicted labels). The value in each box shows the number of cells with the known label for the row, which were predicted to have the column label. The colour corresponds to the proportions of the known label across the different possible predictions; within each row, the colours ‘add up to 1’.

**We can use this to identify groups of cells whose predicted labels differ from their true labels.** For example, considering the cells with true label 6 years (third row from the bottom), two thirds have predicted donor age 5, while the remaining third have predicted donor age 21. [For this example dataset, this analysis doesn’t seem super interesting, but there are others where it is useful! Look at the vignettes for the psupplementary package for more interesting examples.]

```{r fig.height=10,fig.width=10}
plot_predictions_against_classes(sce_best)
```

psupertime can also be applied to data with unknown or different labels. In that case, the x-axis would remain the same, with the labels used to train the psupertime, but the y-axis would be different. Using it on the data used for training means we can check how accurate its labelling is (when psupertime is accurate, all the values should be on the diagonal), and in particular check whether it is less accurate for some labels.


# Session info

```{r}
sessionInfo()
```

