---
title: "Step3.1 Seurat Filter Cells And Preliminary Analysis"
author: "Zhang Yongchao"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    number_sections: TRUE
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      collapse = TRUE,
                      warning = FALSE,
  fig.align = "center")
```

> [Seurat - Guided Clustering Tutorial](see: https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html)

```{r}
Output.Rds.dir <- "./3.1-QC-Passed-Cells-Rds/"
Parameter.Assay.use <- "RNA" # RNA is not supported right now!

if ( !dir.exists(Output.Rds.dir)) {
  dir.create(Output.Rds.dir)
}
```

```{r}
suppressMessages(library(dplyr)) 
suppressMessages(library(Seurat))
suppressMessages(library(patchwork))
suppressMessages(library(ggplot2))
suppressMessages(library(Matrix))
suppressMessages(library(clustree))
suppressMessages(library(scater))
suppressMessages(library(clustree))
suppressMessages(library(Matrix))
suppressMessages(library(DT))
```


# import data

** Attention **: We will also remove cells with zero RNA UMI counts, these cell usually have low CellRanger UMI counts, details see "Q1: what is the zero counts cells" : "Step2.2 RNA QC And Generate Raw Cells Rds"

## import

```{r}
cds.mat <- Read10X(paste0("../../../CellRanger-Outputs/", basename(getwd()),"/outs/filtered_feature_bc_matrix/"))
cds <- CreateSeuratObject(cds.mat,project = basename(getwd()))
DefaultAssay(cds) <- Parameter.Assay.use
cds
```

## Add cell annotations

> [hbctraining - 04_SC_quality_control](https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md)

**number of genes detected per UMI**: this metric with give us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)

```{r}
cds[["percent.mt"]] <- PercentageFeatureSet(cds, pattern = "^mt-",assay = DefaultAssay(cds))
cds[[paste("nCount",DefaultAssay(cds),"RemoveMT",sep = "_")]] <- cds[[paste("nCount", DefaultAssay(cds),sep = "_")]]*(1-cds[["percent.mt"]]/100)
cds[["log10GenesPerUMI"]] <- log10(cds[[paste("nFeature", DefaultAssay(cds),sep = "_")]]) / log10(cds[[paste("nCount", DefaultAssay(cds),sep = "_")]])
```

# QC - pre-filter

```{r fig.width=10,fig.height=12}
qc.keys <- colnames(cds@meta.data)[2:10]
VlnPlot(cds,features = qc.keys, ncol = 3,pt.size = 0.001,group.by = "orig.ident") + NoLegend()
VlnPlot(cds,features = qc.keys, ncol = 3,pt.size = 0,group.by = "orig.ident") + NoLegend()
```

```{r fig.width=11,fig.height=7}
metadata <- cds@meta.data
  
# Visualize the number UMIs/transcripts per cell
p1 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
    geom_vline(xintercept = c(1500,2000,3000,5e4,7e4,1e5)) + NoLegend()

p2 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA_RemoveMT, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
    geom_vline(xintercept = c(1500,2000,3000,5e4,7e4,1e5)) + NoLegend()

# Visualize the distribution of features detected per cell
p3 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nFeature_RNA, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(300,400,500,700,800)) + NoLegend()

# Visualize the distribution of mitochondrial gene expression detected per cell
p4 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(20,30,40,50,60,70)) + NoLegend()


# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
p5 <- metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = c(0.7,0.75,0.8,0.85)) + NoLegend()

wrap_plots(p1,p2,p3,
           p4,p5)
```

```{r fig.width=12,fig.height=4}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
x.line <- c(1000,2000,3000,5e4,1e5)
y.line <- c(300,500,700)

p1 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)

p2 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=log10GenesPerUMI)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)

wrap_plots(p1,p2)


p3 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA_RemoveMT, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)

p4 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA_RemoveMT, y=nFeature_RNA, color=log10GenesPerUMI)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)



wrap_plots(p3,p4)
```

```{r fig.width=12,fig.height=6}
plot1 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),sep = "_"), feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),sep = "_"), feature2 = paste("nFeature", DefaultAssay(cds),sep = "_")) + NoLegend()
plot1 + plot2

plot1 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),"RemoveMT",sep = "_"), feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),"RemoveMT",sep = "_"), feature2 = paste("nFeature", DefaultAssay(cds),sep = "_")) + NoLegend()
plot1 + plot2
```


```{r fig.width=6,fig.height=6}
plotHighestExprs(as.SingleCellExperiment(cds,assay = DefaultAssay(cds)),colour_cells_by = "orig.ident")
```

# Set QC Cut-offs

```{r}
qc.mat <- data.frame(Sample = basename(getwd()),
                     nCount_RNA.min = 3000,
                     nCount_RNA.max = 5e4,
                     nFeature_RNA.min = 1000,
                     nFeature_RNA.max = Inf,
                     percent.mt.max = 10,
                     log10GenesPerUMI.min = 0.8)
qc.mat
```

```{r}
QC.Cut <- function(SeuratObj, QC.mat = qc.mat){
  Sample.meta <- SeuratObj@meta.data[,c("nCount_RNA","nFeature_RNA","percent.mt","log10GenesPerUMI")]
  dplyr::mutate(Sample.meta,
                nCount_RNA = nCount_RNA > QC.mat$nCount_RNA.min & nCount_RNA < QC.mat$nCount_RNA.max,
                nFeature_RNA = nFeature_RNA > QC.mat$nFeature_RNA.min & nFeature_RNA < QC.mat$nFeature_RNA.max,
                percent.mt = percent.mt < QC.mat$percent.mt.max,
                log10GenesPerUMI = log10GenesPerUMI > QC.mat$log10GenesPerUMI.min)
}
sample.meta <- QC.Cut(cds)
sample.statics <- apply(!sample.meta, 2, sum)
names(sample.statics) <- paste("RemovedBy",names(sample.statics),sep = "_")
statics.df <- data.frame(cell.all = length(colnames(cds)), t(sample.statics), cell.left = sum(apply(sample.meta, 1, all)))
qc_ststics.file <- paste(Output.Rds.dir,"qc.statics.csv",sep = "")
qc_ststics <- cbind(qc.mat,statics.df)
qc_ststics
write.csv(qc_ststics,file = qc_ststics.file)
```


# QC - post filter

```{r}
cells.kept <- colnames(cds)[unname(apply(sample.meta, 1, all))]
cds <- subset(cds, cells = cells.kept)
cds
```

```{r fig.width=10,fig.height=12}
qc.keys <- colnames(cds@meta.data)[2:10]
VlnPlot(cds,features = qc.keys, ncol = 3,pt.size = 0.001,group.by = "orig.ident") + NoLegend()
VlnPlot(cds,features = qc.keys, ncol = 3,pt.size = 0,group.by = "orig.ident") + NoLegend()
```

```{r fig.width=11,fig.height=7}
metadata <- cds@meta.data
  
# Visualize the number UMIs/transcripts per cell
p1 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
    geom_vline(xintercept = c(1500,2000,3000,5e4,7e4,1e5)) + NoLegend()

p2 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA_RemoveMT, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
    geom_vline(xintercept = c(1500,2000,3000,5e4,7e4,1e5)) + NoLegend()

# Visualize the distribution of features detected per cell
p3 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nFeature_RNA, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(300,400,500,700,800)) + NoLegend()



# Visualize the distribution of mitochondrial gene expression detected per cell
p4 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = c(20,30,40,50,60,70)) + NoLegend()


# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
p5 <- metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = c(0.7,0.75,0.8,0.85)) + NoLegend()


wrap_plots(p1,p2,p3,
           p4,p5)
```

```{r fig.width=12,fig.height=8}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
x.line <- c(1000,2000,3000,5e4,1e5)
y.line <- c(300,500,700)

p1 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)

p2 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=log10GenesPerUMI)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)


wrap_plots(p1,p2)


p5 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA_RemoveMT, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)

p6 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA_RemoveMT, y=nFeature_RNA, color=log10GenesPerUMI)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = x.line) +
  	geom_hline(yintercept = y.line)

wrap_plots(p5,p6)
```

```{r fig.width=12,fig.height=6}
plot1 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),sep = "_"), feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),sep = "_"), feature2 = paste("nFeature", DefaultAssay(cds),sep = "_")) + NoLegend()
plot1 + plot2

plot1 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),"RemoveMT",sep = "_"), feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(cds, feature1 = paste("nCount", DefaultAssay(cds),"RemoveMT",sep = "_"), feature2 = paste("nFeature", DefaultAssay(cds),sep = "_")) + NoLegend()
plot1 + plot2
```


```{r fig.width=6,fig.height=6}
plotHighestExprs(as.SingleCellExperiment(cds,assay = DefaultAssay(cds)),colour_cells_by = "orig.ident")
```



# seurat clustering analysis

```{r}
s.genes <- readRDS("~/reference/Biology-Database/Cell-Cycles-Genes/Mouse/mouse.s.genes.rds")
g2m.genes <- readRDS("~/reference/Biology-Database/Cell-Cycles-Genes/Mouse/mouse.g2m.genes.rds")
s.genes
g2m.genes

cds <- NormalizeData(cds,assay = DefaultAssay(cds))
cds <- FindVariableFeatures(cds, selection.method = "vst", nfeatures = 3000,assay = DefaultAssay(cds))
# Remove mito genes from variable gnes
VariableFeatures(cds) <- grep("^mt-",VariableFeatures(cds),value = TRUE,invert = TRUE)

cds <- CellCycleScoring(cds, s.features = s.genes, g2m.features = g2m.genes,assay = DefaultAssay(cds))
cds <- ScaleData(cds, vars.to.regress = c("percent.mt",paste("nCount", DefaultAssay(cds),sep = "_"),"S.Score", "G2M.Score"),assay = DefaultAssay(cds))
cds <- RunPCA(cds, features = VariableFeatures(cds),assay = DefaultAssay(cds))
cds <- RunUMAP(cds, dims = 1:20,assay = DefaultAssay(cds))
cds <- RunTSNE(cds, dims = 1:20,assay = DefaultAssay(cds))
cds <- FindNeighbors(cds, dims = 1:20, assay = DefaultAssay(cds)) %>% FindClusters(resolution = c(0.2,0.4,0.6,0.8,1.0,1.2,1.4,1.6), assay = DefaultAssay(cds))
```

# Cell numbers in each resolution

```{r}
statics.cluster.cells <- function(seurat){
  test.data <- as.data.frame(as.matrix(seurat@meta.data[,grepl("(_snn_res)|(^Annotation)",colnames(seurat@meta.data))]),stringsAsFactors = FALSE)
  test.data$cell <- rownames(test.data)
  rownames(test.data) <- NULL
  test.data.melt <- reshape2::melt(test.data, id.vars = "cell",value.name = "cluster")
  test.data.melt.dcast <- reshape2::dcast(test.data.melt,formula = cluster ~ variable,fun.aggregate = length)
  test.data.melt.dcast$cluster <- as.numeric(test.data.melt.dcast$cluster)
  test.data.melt.dcast <- arrange(test.data.melt.dcast,cluster)
  rownames(test.data.melt.dcast) <- test.data.melt.dcast$cluster
  test.data.melt.dcast$cluster <- NULL
  return(test.data.melt.dcast)
}
DT::datatable(statics.cluster.cells(cds),  extensions = c('Buttons','ColReorder','FixedColumns'),
              options = list(pageLength = 30, 
                             autoWidth = TRUE,
                             scrollX = TRUE,
                             scrollY = FALSE,
                             dom = 'Bfrtip', # Buttons | note B in dom denotes the place where the buttons are to be inserted
                             buttons = c('csv', 'pdf'),
                             colReorder = TRUE,
                             fixedColumns = TRUE)
              )
```

# Choose Resolution

```{r fig.height=10,fig.width=10}
clustree(cds@meta.data,prefix = paste(DefaultAssay(cds),"_snn_res.",sep = ""))
```

```{r fig.width=6,fig.height=5}
Idents(cds) <- paste(DefaultAssay(cds),"_snn_res.0.6",sep = "")
DimPlot(cds,label = TRUE)
DimPlot(cds,label = TRUE,reduction = "tsne")

DimPlot(cds,group.by = "Phase")
DimPlot(cds,reduction = "tsne",group.by = "Phase")
```

# Cluster QC

```{r fig.width=12,fig.height=8}
VlnPlot(cds,features = qc.keys, pt.size = 0, ncol = 3) + NoLegend()
```

```{r fig.width=17,fig.height=12}
FeaturePlot(cds,features = c("Pdgfra","Epcam","Ptprc","Vwf","Pdgfrb","Nrxn1",'Rgs5','Acta2','Cd34','Il33','Il1rl1')) & NoLegend()
```

```{r fig.width=16,fig.height=14}
FeaturePlot(cds,features = qc.keys)
```

# save Rds and mtx file for scrublet doublet analysis

```{r}
saveRDS(cds,file = paste(Output.Rds.dir,basename(getwd()),".rds",sep = ""))

Amatrix <- as.matrix(cds@assays$RNA@counts)
writeMM(Matrix(Amatrix,sparse = TRUE),file = paste0(Output.Rds.dir,"UMI.mtx"))
```

# Session info

```{r}
sessionInfo()
```